NSAttributedString: Change stringEncoding from NSUInteger to unsigned long

LumioseSil opened the pull request:
#### 0c8afd4d26dffc9006120ee0d11cb862a63b3873
<pre>
NSAttributedString: Change stringEncoding from NSUInteger to unsigned long
<a href="https://bugs.webkit.org/show_bug.cgi?id=251687">https://bugs.webkit.org/show_bug.cgi?id=251687</a>

Reviewed by NOBODY (OOPS!).

CFStringConvertEncodingToIANACharSetName takes an unsigned long,
not an NSUInteger, which matters on ARM64_32 systems like watchOS,
so stringEncoding should be an unsigned long variable.

* Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
  (loadFromHTMLWithData:options:completionHandler): Changed
  stringEncoding from NSUInteger to unsigned long.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c8afd4d26dffc9006120ee0d11cb862a63b3873

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/113273 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22406 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1945 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/5046 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121727 "Built successfully") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23774 "Built successfully") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/6265 "Build is in progress. Recent messages:Failed to print configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; Running layout-tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/119060 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100930 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106371 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1470 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46721 "Build is in progress. Recent messages:Failed to print configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 19 flakes 252 failures 1 missing results; Uploaded test results; 30 flakes 163 failures 1 missing results; compiling; Running layout-tests-repeat-failures-without-change") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14706 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1509 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/98974 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15414 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20708 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53500 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/17269 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -456,11 +456,11 @@ + (void)loadFromHTMLWithData:(NSData *)data options:(NSDictionary<NSAttributedSt
         auto* textEncodingName = dynamic_objc_cast<NSString>(options[NSTextEncodingNameDocumentOption]);
-        auto characterEncoding = static_cast<NSStringEncoding>(dynamic_objc_cast<NSNumber>(options[NSCharacterEncodingDocumentOption]).unsignedIntegerValue);
         auto* baseURL = dynamic_objc_cast<NSURL>(options[NSBaseURLDocumentOption]);
+        unsigned long characterEncoding = dynamic_objc_cast<NSNumber>(options[NSCharacterEncodingDocumentOption]).unsignedLongValue;
```
This change seems incorrect. Documentation at https://developer.apple.com/documentation/uikit/nscharacterencodingdocumentoption?language=objc says "NSNumber containing the unsigned int NSStringEncoding".

AZero13 replied:
<img width="827" alt="Screen Shot 2023-02-23 at 4 52 58 PM" src="https://user-images.githubusercontent.com/83477269/221038854-6a9f0ca3-3b85-4645-8ccf-464e6ead0c76.png">
RequestResourceCocoa, the other usage of the function, has the unsigned long value. So, is that wrong too? But the CFStringConvertNSStringEncodingToEncoding function takes an unsigned long

aproskuryakov replied:
I haven't analyzed RequestResourceCocoa code, perhaps it's slightly wrong, or perhaps `-contentDispositionEncodingFallbackArray` returns unsigned longs.

This updated PR still appears to be incorrect, because NSCharacterEncodingDocumentOption is NOT an unsigned long type, but an unsigned int.

AZero13 replied:
NSCharacterEncodingDocumentOption is an NSUInteger, which is an unsigned long on 64 bits, unsigned int in 32 bits. CFStringConvertNSStringEncodingToEncoding explicitly always takes an unsigned long, not NSUInteger.

This is why the variable type has been changed to an unsigned long by the way, because it is not NSStringEncoding in the function signature.

aproskuryakov replied:
It is not clear to me what you are getting at. I don't even see where the code cleanup is, much less any bug fix.

If you think that this PR is necessary, please clearly explain why.

aproskuryakov commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -456,11 +456,11 @@ + (void)loadFromHTMLWithData:(NSData *)data options:(NSDictionary<NSAttributedSt
 
         if (characterEncoding && !textEncodingName) {
-            auto stringEncoding = CFStringConvertNSStringEncodingToEncoding(characterEncoding);
+            CFStringEncoding stringEncoding = CFStringConvertNSStringEncodingToEncoding(characterEncoding);
```
And then CFStringConvertNSStringEncodingToEncoding takes an unsigned long for some reason, but an unsigned int fits in this type, so no change seems to be necessary.

AZero13 replied:
The other example of this function being called in WebKit does have the unsigned long explicitly, so that's inconsistent with this. 

But this is about the arguments to the IANAChar function.

aproskuryakov replied:
> But this is about the arguments to the IANAChar function.

CFStringConvertEncodingToIANACharSetName takes CFStringEncoding, which is what we pass with or without this PR. It just changes the variable declaration to not use auto, but the actual type remains the same.

AZero13 replied:
Documention says one thing, but the function signature for CFStringConvertNSStringEncodingToEncoding is another. I am following the signature, because that takes an unsigned long explicitly, not an NSUInteger.

LumioseSil committed:
NSAttributedString: Change stringEncoding from NSUInteger to unsigned long
https://bugs.webkit.org/show_bug.cgi?id=251687

Reviewed by NOBODY (OOPS!).

CFStringConvertEncodingToIANACharSetName takes an unsigned long,
not an NSUInteger, which matters on ARM64_32 systems like watchOS,
so stringEncoding should be an unsigned long variable.

* Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
  (loadFromHTMLWithData:options:completionHandler): Changed
  stringEncoding from NSUInteger to unsigned long.
